crates/loader/src/validation.rs

```diff
@@ -1,23 +1,276 @@
 #![deny(missing_docs)]
 
-use anyhow::{Context, Result};
+use std::str::FromStr;
+
+use anyhow::{anyhow, Result};
 use reqwest::Url;
+use spin_manifest::{AllowedHttpHost, AllowedHttpHosts};
 
-// Check whether http host can be parsed by Url
+// Checks a list of allowed HTTP hosts is valid
 pub fn validate_allowed_http_hosts(http_hosts: &Option<Vec<String>>) -> Result<()> {
-    if let Some(domains) = http_hosts.as_deref() {
-        if domains
-            .iter()
-            .any(|domain| domain == wasi_outbound_http::ALLOW_ALL_HOSTS)
-        {
-            return Ok(());
+    parse_allowed_http_hosts(http_hosts).map(|_| ())
+}
+
+// Parses a list of allowed HTTP hosts
+pub fn parse_allowed_http_hosts(raw: &Option<Vec<String>>) -> Result<AllowedHttpHosts> {
+    match raw {
+        None => Ok(AllowedHttpHosts::AllowSpecific(vec![])),
+        Some(list) => {
+            if list
+                .iter()
+                .any(|domain| domain == wasi_outbound_http::ALLOW_ALL_HOSTS)
+            {
+                Ok(AllowedHttpHosts::AllowAll)
+            } else {
+                let parse_results = list
+                    .iter()
+                    .map(|h| parse_allowed_http_host(h))
+                    .collect::<Vec<_>>();
+                let (hosts, errors) = partition_results(parse_results);
+
+                if errors.is_empty() {
+                    Ok(AllowedHttpHosts::AllowSpecific(hosts))
+                } else {
+                    Err(anyhow!(
+                        "One or more allowed_http_hosts entries was invalid:\n{}",
+                        errors.join("\n")
+                    ))
+                }
+            }
+        }
+    }
+}
+
+fn parse_allowed_http_host(text: &str) -> Result<AllowedHttpHost, String> {
```

Comment:
Would this make sense as `impl FromStr for AllowedHttpHost`?

Reply:
It would, but that would mean putting the parse logic into the manifest crate, which is generally concerned with execution-ready forms rather than stringised formats in the TOML file.  The loader crate is, I think, the right place for this code, but it can't add new trait implementations.

Reply:
I guess I'll be updating this as part of the spin-core migration anyway, which should move this logic into `outbound-http`.